src/hotspot/share/gc/g1/g1FullGCPrepareTask.cpp

```diff
@@ -61,13 +62,25 @@ bool G1FullGCPrepareTask::G1CalculatePointersClosure::do_heap_region(HeapRegion*
     }
   } else {
     assert(!hr->is_humongous(), "moving humongous objects not supported.");
-    prepare_for_compaction(hr);
+    size_t live_words = _collector->hr_live_words(hr->hrm_index());
+    size_t live_words_threshold = _collector->scope()->hr_live_words_threshold();
+
+    if(live_words <= live_words_threshold) {
+      prepare_for_compaction(hr);
+    } else {
+      assert(MarkSweepDeadRatio > 0,
+                "it should not trigger skipping compaction, when MarkSweepDeadRatio == 0");
+      // Force the high live ration region pinned,
+      // as we need skip these regions in the later compact step.
+      force_pinned = true;
+      log_debug(gc, phases)("Phase 2: skip compaction region index: %u, live words: " SIZE_FORMAT,
+                            hr->hrm_index(), live_words);
+    }
   }
 
   // Reset data structures not valid after Full GC.
   reset_region_metadata(hr);
-
-  _collector->update_attribute_table(hr);
+  _collector->update_attribute_table(hr, force_pinned);
```

Comment:
@tschatzl, this call was here since before (but without the force part). Why is it needed? The call to `update_attribute_table()` will return straight away if the region has been free and to me it looks like that is the only change that can have occurred at this point. If it is not needed for other reasons, instead of adding the `force_pinned` part we could just add a `set_pinned()` helper to the `G1FullCollector` to make it more clear what is happening.

Reply:
in G1ResetPinnedClosure,  we need to skip freed humongous regions which depends on the info in attribute table. But if we don't update the corresponding table entry for freed humongous regions, these regions will invoke reset_pinned_after_full_gc(), seems it's wrong.

Reply:
Yes, and previously this has been done by checking `r->is_pinned()` so it didn't use the information from the attribute table and you are correct that now when using it we need to make sure freed humongous are skipped. To me it looks like they would not be skipped right now. Will start some testing to see if we can trigger this.

Reply:
Realize now that the new code does things for free regions. 👍